content/blog/gatbsy-typescript/index.md

```diff
@@ -0,0 +1,284 @@
+it will ask for props definitions and perhaps some dependencies type definitions.
+
+### `declarations.d.ts`
+
+In most Typescript codebases we have some global types or some untyped modules to declare,
+that is why you most likely will need `src/declaration.d.ts`. Let's see a bair minimum one:
+
+```typescript
+// Make typescript aware of this
+// global Gatbsy variable
+declare var __PATH_PREFIX__: string
+
+// Declare that this modules are untyped
+declare module "typography-theme-github"
+declare module "vfile-message"
+// This makes importing svg images
+// typecheck. You might need
+// to declare other types of files
+// used in your codebase
+declare module "*.svg"
+```
+
+And then you should import this file from your `gatsby-browser.js` which
+is the main entry point of the browser application:
+
+```javascript
+import "./src/declarations.d.ts"
+```
+
+### Typechecking
+
+As the typescript plugin docs state, by default it wont run typechecks on your code, it
+will only strip out any type information and typescript sintax and compile it down to Javascript.
+
+For enabling typechecking you will need to create an npm script:
+
+```json
+"typecheck": "tsc --noEmit",
+```
+
+`tsc` is the Typescript Compiler and `noEmit` tells it that we only want to typecheck.
+
+I also call this script from my `test` script because I want to always pass this verification,
+which as one of the reasons I had to add it.
+
+You will probably need to install type definitions for some libraries you might be using by
+`yarn add --dev @types/my-lib` if they exist or simply declaring them as untyped.
+When typechecking your code base the typescript compiler will let you know what to do
+pretty acurately in each case.
+
+**NOTE**: I configured this in March 2020 and I had some troubles with Gatbsy type definitions, in
+order to fix it I had to add this to my `package.json#resolutions` field:
+
+```json
+"resolutions": {
+  "unified": "7.1.0",
+  "vfile": "4.0.3",
+  "@types/vfile": "4.0.0"
+},
+```
+
+If you see a type error relating these libraries this might be fix for you.
+
+### gatsby-plugin-typescript
+
+I did not made any special configurations to this plugin.
+
+### gatsby-plugin-graphql-codegen
+
+Here is where things become more complicated.
+
+The main data layer of Gatbsy is Grapqhl and most of the pages
+and some components will acess that data layer and that data will
+flow through props and context down the component tree.
+
+So it is reasonable to expect to somehow have thata data model
+and the queries we make out of it to by statically typed.
+
+You can of course write these types by hand but it will be much better
+to have a tool automatically generate those files from the latest queries
+and GraphQL data model. Enter: `gatsby-plugin-graphql-codegen`.
+
+This is how I recommend you configure it:
+
+```typescript
+{
+  resolve: `gatsby-plugin-graphql-codegen`,
+  options: {
+    codegenConfig: {
+      // optional if you prefix, as I do,
```

Comment:
explain why it could be convenient maybe? 
I recall seeing that and being unsure regarding why it would be that way when it was first implemented (I still don't really know)

Reply:
done